### ISSUE TITLE
[Bug Report][Android] waitUntilFirstFrameRendered completes too early

### ISSUE THREAD

**Issue by clragon** [media-kit/media-kit#349]
When using `waitUntilFirstFrameRendered` to show a thumbnail instead of the `Video` widget, we get the desired behaviour on Windows.
However, on Android devices, larger videos can cause a period where `waitUntilFirstFrameRendered` has completed and we switch to the `Video` widget, only to find that the first frame has not actually rendered. This then causes a less than optimal user experience where a black screen is displayed for a duration.
This behaviour is not reproducible on Windows. I assume this is because the performance on Windows simply eradicates this problem. It also does not occur with short videos on Android. 

I provide the following test code to reproduce this behaviour:

<details>
  <summary>Source Code</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:media_kit/media_kit.dart';
import 'package:media_kit_video/media_kit_video.dart';
import 'package:rxdart/rxdart.dart';

class VideoPlayer extends Player {
  VideoPlayer() {
    controller.waitUntilFirstFrameRendered.then((_) => _initialized.add(true));
  }

  late final VideoController _controller = VideoController(this);
  VideoController get controller => _controller;

  final BehaviorSubject<bool> _initialized = BehaviorSubject.seeded(false);
  Stream<bool> get initialized => _initialized.stream;

  bool get isInitialized => _initialized.value;
}

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  MediaKit.ensureInitialized();
  runApp(const App());
}

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  Stream<bool>? initialized;
  VideoPlayer? player;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('waitUntilFirstFrameRendered test'),
      ),
      body: Builder(
        builder: (context) {
          if (initialized == null) {
            return Center(
              child: TextButton(
                child: const Text('Start'),
                onPressed: () => setState(() {
                  player = VideoPlayer();
                  player!.open(
                    Media(
                      'https://jsoncompare.org/LearningContainer/SampleFiles/Video/MP4/Sample-Video-File-For-Testing.mp4',
                    ),
                    play: false,
                  );
                  initialized = player!.initialized;
                  player!.play();
                }),
              ),
            );
          }
          return Center(
            child: AspectRatio(
              aspectRatio: 16 / 9,
              child: StreamBuilder<bool>(
                stream: initialized,
                builder: (context, snapshot) {
                  bool initialized =
                      snapshot.data ?? (player?.isInitialized ?? false);
                  if (!initialized) {
                    return Container(
                      color: Colors.red,
                    );
                  }
                  return Video(
                    controller: player!.controller,
                    fill: Colors.transparent,
                  );
                },
              ),
            ),
          );
        },
      ),
    );
  }
}
```

</details>

Pressing start will show a red Container while we wait for `waitUntilFirstFrameRendered`. As soon as it completes, we show the `Video` instead. On Android, we can see a black screen between showing our red Container and actually playing the Video.

The duration of the black screen seems to vary based on the video. I have tried to choose a sample video that demonstrates this behaviour well enough.

**Comment by alexmercerind** [comment]
Thanks for the descriptive bug-report & example.

It is indeed possible. We do not run our own render-loop on Android (unlike other platforms), this brings a challenge to identify first frame render. There should be some solution, I will try to address it.

**Comment by alexmercerind** [comment]
I've been trying to play with the API, but I always have the first frame rendered as intended.

Theoretically, there is indeed a possibility of a black screen for a fraction of time (since we don't have our own render loop on Android to get notified from), but still, it won't classify as "too early".

https://github.com/media-kit/media-kit/assets/28951144/1f3fa3a4-275b-4d86-83f8-25f701eadd28

https://github.com/media-kit/media-kit/assets/28951144/1159122b-d164-4f62-8e2e-a9f1e98638a2

<hr>

Looking for any possibilities...

**Comment by clragon** [comment]
I see. Its unfortunate that you cannot reproduce it. Maybe there are hardware differences?

I will provide a video of my own:

https://github.com/media-kit/media-kit/assets/11785085/8ada9630-34a2-4f07-9f82-d2bfd7ce5407

Here we can see the blackscreen before the video is actually shown. 
On other videos, this can show up even longer, up to several seconds. 

My device information is as follows:

- Device: Samsung Galaxy S20 FE 5G (SM-G781B/DS)
- OS: Android 13

Please let me know if you need additional information regarding the issue.

**Comment by alexmercerind** [comment]
A new update has been released.

Please refer to the "Installation" section of the README.

**Comment by clragon** [comment]
thank you @alexmercerind
On the latest release however, 

```yml
  media_kit: ^1.1.4
  media_kit_video: ^1.1.4
  media_kit_native_event_loop: ^1.0.7
  media_kit_libs_android_video: ^1.3.2
  media_kit_libs_ios_video: ^1.1.3
  media_kit_libs_macos_video: ^1.1.3
  media_kit_libs_windows_video: ^1.0.7
  media_kit_libs_linux: ^1.1.1
```

I no longer get any video whatsoever on android.
The red container never disappears; The video is playing however, as I can hear audio if there is any. 
It seems in this case the Future no longer completes at all, or completes and then un-completes immediately.
Other platforms work fine as usual.

**Comment by alexmercerind** [comment]
Oh man...

**Comment by alexmercerind** [comment]
Seems like some symbols etc. get stripped in release mode... Looking for some solution.

**Comment by alexmercerind** [comment]
Hi @clragon,

media_kit_video: ^1.1.5 should do it.

**Comment by alexmercerind** [comment]
Few more recommendations which I wanted to provide:

<hr>

1. You might wanna "intercept" `waitUntilFirstFrameRendered` if `Player.stream.error` is fired.

This should be done to avoid any possible dead-lock. For instance, you keep waiting for the first frame render, which never happens because of some network issue or missing URL etc.

<hr>

2. Please [enable --split-per-abi](https://docs.flutter.dev/deployment/android#what-is-a-fat-apk) or [use app bundle (instead of APK)](https://docs.flutter.dev/deployment/android#when-should-i-build-app-bundles-versus-apks) on Android.

Distributing Android applications that contain native (e.g. C/C++) code without --split-per-abi or app bundle causes size to be substantially higher; about 3-4 ×. In case of native libraries, shared libraries (`*.so`) for each architecture (arm64, armeabi, x86_64, x86) are bundled separately, this multiplies the resulting size for a fat APK (i.e. APK containing all architectures). Java libraries on the other hand (unlike native libraries), run inside JVM & a single "copy" (I don't know how Java compiles) of instructions can be used by all architectures. 

The Flutter engine is C/C++ i.e. native, Dart code is compiled to native, package:media_kit is native. All are bundled separately inside APK to support all 3 architectures that Flutter supports (arm64, armeabi, x86_64), this is not elegant.

Infact: package:media_kit also provides x86 shared libraries in order to support emulators on AMD machines (where only x86 works). While Flutter itself does not support x86 in release mode, this makes bundling of those x86 shared libraries in release-mode APK completely redundant.

**Comment by clragon** [comment]
I have now spent several hours trying to figure this out but I am met with utter confusion;

1. My sample code from above, when inserted into the `media_kit_test` repository works as expected
2. My sample code from above, when inserted in my target application, resolves `waitUntilFirstFrameRendered` within ~10 milliseconds, regardless whether the video is initialized or not
3. My sample code from above, when inserted into a freshly generated flutter project _never_ resolves `waitUntilFirstFrameRendered` and instead loads indefinitely.

For testing:

- I have replaced the entire code of my target application with just the sample code with no success.
- I have entirely deleted and replaced the `android` folder of `media_kit_test` with a freshly generated one but I have not spotted any notable differences in the git diff.
- I have replaced my dependencies in the freshly generated flutter project with direct git URLs instead of versions with no success.

I cannot figure out why these behaviours differ. Why does the correct behaviour only show in `media_kit_test`?
All the APKs were tested without any split compilation and on the same device I have mentioned earlier.

For reproduction:

- My [target application](https://github.com/clragon/e1547).
- The fresh flutter project was generated on Flutter `3.10.6` stable, Dart `3.0.6`.

In both cases I have replaced the entire `main.dart` file with the sample code provided originally.

What could be the difference between these projects? 
From what I can tell its neither the android folder, nor the dart code, nor the version being published incorrectly.

**Comment by alexmercerind** [comment]
https://github.com/media-kit/media-kit/commit/b7cca846acee9e5748b1ccc0f8c3226057aec6fd

<hr>

Apologies for the delay.

I truly appreciate your detailed bug report & patience.

Since it wasn't really feasible to get notified about the new frame render in our implementation, I'm using some private Flutter API through reflection.

I have added a fallback case where `waitUntilFirstFrameRendered` will be automatically resolved (to avoid deadlock) if that private API could not be located (most likely due to changes in the Flutter engine). I don't expect embedding code to be changed much inside the Flutter engine. Obviously, if things change in any future Flutter update, I will migrate). You were hitting this fallback because I didn't know there's a `FlutterFragmentActivity`.

All the cases:

- media_kit_test: Uses [`FlutterActivity`](https://github.com/media-kit/media-kit/blob/d4100553068b56c48d8ab1eaa0f19620fc01f796/media_kit_test/android/app/src/main/kotlin/com/alexmercerind/media_kit_test/MainActivity.kt#L3).
  Already works!
- e1547: Uses [`FlutterFragmentActivity`](https://github.com/clragon/e1547/blob/444ca82970cd3b54959278339f1cb6ec322d380d/android/app/src/main/kotlin/net/e1547/MainActivity.kt#L3).
  Resolved!
- New project: Probably you missed `android.permission.INTERNET` in release-mode, I can't reproduce.
  Again, a good idea will be to intercept your `BehaviorSubject` based on `Player.stream.error` as well.


`waitUntilFirstFrameRendered` on Android... has been some journey.

We will land #383 & close this issue after a stable release. Thanks!

**Comment by clragon** [comment]
Thank you very much.

Would it be possible for `waitUntilFirstFrameRendered` to throw if the Player encounters an error?
This would make handling the deadlocks a bit easier, I believe.

**Comment by alexmercerind** [comment]
Hi! I published an update media_kit: ^1.1.5 & media_kit_video: ^1.1.6.

They should do it for you.

**Comment by deckerst** [comment]
I confirm the fix works for me with the latest version (I had the same issue as I use `FlutterFragmentActivity`). Thanks!
